hledger-lib/Hledger/Reports/ReportOptions.hs

```diff
@@ -719,227 +692,95 @@ setWithReport l = overWithReport l . const
 type ReportableLens' s a = forall f. Reportable f String => (a -> f a) -> s -> f s
 
 -- | Lenses for ReportOpts.
--- Note that some of these are not true lenses, as they have a further restriction on
+
+-- Implement HasReportOptsNoUpdate, the basic lenses for ReportOpts.
+makeHledgerClassyLenses ''ReportOpts
+makeHledgerClassyLenses ''ReportSpec
+
+-- | Special lenses for ReportOpts which also update the Query and QueryOpts in ReportSpec.
+-- Note that these are not true lenses, as they have a further restriction on
 -- the functor. This will work as a normal lens for all common uses, but since they
 -- don't obey the lens laws for some fancy cases, they may fail in some exotic circumstances.
--- The special type is so that updating querystring_, real_, depth_, date2_, period_, or
--- statuses_ can automatically update the Query and QueryOpts in ReportSpec.
-class HasReportOpts a where
+--
+-- Note that setWithReport/overWithReport should only be necessary with
+-- querystring and reportOpts: the other lenses should never fail.
+--
+-- === Examples:
+-- >>> import Lens.Micro (set)
+-- >>> _rsQuery <$> setWithReport querystring ["assets"] defreportspec
+-- Right (Acct (RegexpCI "assets"))
+-- >>> _rsQuery <$> setWithReport querystring ["(assets"] defreportspec
+-- Left "this regular expression could not be compiled: (assets"
+-- >>> _rsQuery $ set querystring ["assets"] defreportspec
+-- Acct (RegexpCI "assets")
+-- >>> _rsQuery $ set querystring ["(assets"] defreportspec
+-- *** Exception: Updating ReportSpec failed: try using overWithReport instead of over or setWithReport instead of set
+-- >>> _rsQuery $ set period (MonthPeriod 2021 08) defreportspec
+-- Date DateSpan 2021-08
+class HasReportOptsNoUpdate a => HasReportOpts a where
     reportOpts :: ReportableLens' a ReportOpts
-    reportOptsNoUpdate :: Lens' a ReportOpts
+    reportOpts = reportOptsNoUpdate
+    {-# INLINE reportOpts #-}
```

Comment:
Maybe I asked this before: why the INLINE pragmas here, do we know they are worthwhile ?

Reply:
I put the INLINE pragmas in because the Template Haskell splices put them in. I'm not sure whether they are worthwhile, but they are also low-cost. I could remove them if you'd prefer.

Reply:
Whatever you prefer, but I'd probably leave them out for now, on the basis that GHC quite likely can figure this out by itself, these probably aren't called in a hot loop, and we don't want to have to be annotating things if we can avoid it.

Reply:
Edward Kmett (author of the lens library) on using INLINE pragmas in lens instance declarations.
https://www.reddit.com/r/haskell/comments/cjkc3l/should_i_be_inlining_instance_implementations/

Based on this I'm inclined to leave it in.